### PR TITLE
Admin UI

### PR DESCRIPTION
**Pull request by tekrajchhetri**
This pull request implements the following:

1. Integration with Google/Github OAuth for login.
2. Additionally, some basic UI skeleton (WOP) for both the admin + user side.
3. UI configuration based on config file (See https://sensein.group/brainkbdocs/deployment_userinterface.html)
4. Knowledge base page: Pagination + integration to query service 
5. Skeleton page for data release.

### CODE REVIEW

File: djangoUImanager/BrainKBUIManager/settings.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+# SECURITY WARNING: keep the secret key used in production secret!
+SECRET_KEY = "django-insecure-)@-kr16hschn*ezk-=_tfg(pomnf2e*qk&p#avw+^=y-t0ibsu"
+
+# SECURITY WARNING: don't run with debug turned on in production!
+DEBUG = True
+CORS_ALLOWED_ORIGINS = [
+    'http://localhost:3000',
+    'http://127.0.0.1:3000',
+]
```
Comment (aaronkanzer) [inline djangoUImanager/BrainKBUIManager/settings.py:30]: You'll need to bump these once you get past local dev

File: djangoUImanager/BrainKBUIManager/settings.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+BASE_DIR = Path(__file__).resolve().parent.parent
+
+# Quick-start development settings - unsuitable for production
+# See https://docs.djangoproject.com/en/4.2/howto/deployment/checklist/
+
+# SECURITY WARNING: keep the secret key used in production secret!
+SECRET_KEY = "django-insecure-)@-kr16hschn*ezk-=_tfg(pomnf2e*qk&p#avw+^=y-t0ibsu"
+
+# SECURITY WARNING: don't run with debug turned on in production!
+DEBUG = True
```
Comment (aaronkanzer) [inline djangoUImanager/BrainKBUIManager/settings.py:26]: I'd make a separate dev settings file so that this doesnt leak into other environments

File: djangoUImanager/README.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# BrainKB UI
```
Comment (aaronkanzer) [inline djangoUImanager/README.md:1]: Perhaps a bit of cleanup for this file? It doesn't seem to instruct much

File: djangoUImanager/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+WORKDIR .
+
+# Install dependencies
+RUN pip install -r requirements.txt
+# Run migrations and create superuser
+RUN python manage.py makemigrations
+RUN python manage.py migrate
+#RUN DJANGO_SUPERUSER_USERNAME=brainadmin1 \
+#DJANGO_SUPERUSER_PASSWORD=brainadmin1 \
+#DJANGO_SUPERUSER_EMAIL="brainadmin@admin.com" \
```
Comment (aaronkanzer) [inline djangoUImanager/Dockerfile:15]: Were these files intended to be commented out?

File: djangoUImanager/Web/migrations/0001_initial.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+# Generated by Django 3.2.18 on 2024-04-25 13:12
```
Comment (aaronkanzer) [inline djangoUImanager/Web/migrations/0001_initial.py:1]: You should be able to flatten your migrations here into one (seems you have 13 here)

File: djangoUImanager/Web/models.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    query_url = models.URLField(blank=False, unique=True)
+    query_endpoint_type = models.CharField(max_length=20, unique=True, choices=ENDPOINT_TYPE, default='get')
+    endpoint_service_type = models.CharField(max_length=20, choices=ENDPOINT_SERVICE_TYPE, default='query')
+    status_active = models.BooleanField(default=False)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+
+class UserProfile(models.Model):
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
```
Comment (aaronkanzer) [inline djangoUImanager/Web/models.py:57]: Is there a reason you declare an `id` field here and not elsewhere?

File: djangoUImanager/Web/models.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    orcid_id = models.CharField(blank=True, max_length=350)
+    role = models.CharField(max_length=30, choices=ROLES_BRAINKB, default='curator')
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    def picture(self):  # new
+        return mark_safe(f'<img src = "{self.profile_pic.url}" width = "100"/>')
+
+class Comment(models.Model):
+    user_profile = models.ForeignKey(UserProfile, on_delete=models.CASCADE, related_name='comments')
```
Comment (aaronkanzer) [inline djangoUImanager/Web/models.py:76]: Are you sure you want `on_delete=models.CASCADE`?

File: djangoUImanager/Web/models.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    comment_text = models.TextField()
+    commented_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+class Revision(models.Model):
+    user_profile = models.ForeignKey(UserProfile, on_delete=models.CASCADE, related_name='revisions')
```
Comment (aaronkanzer) [inline djangoUImanager/Web/models.py:82]: Are you sure you want `on_delete=models.CASCADE`?

File: djangoUImanager/api/urls.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+# @Web     : https://tekrajchhetri.com/
+# @File    : urls.py
+# @Software: PyCharm
+
+
+from django.urls import path, include
+from . import views
+
+
+urlpatterns = [
```
Comment (aaronkanzer) [inline djangoUImanager/api/urls.py:24]: Your API design seems to be a bit distributed in regards to URLs -- e.g. you have a separate URL entirely for GET, PATCH, POST, etc. HTTP requests

Perhaps think about the API design as:

`/api/knowledgebase/` as the root, and then, depending on what query arguments, or HTTP request, a certain action happens.

e.g.

`/api/knowledgebase?knowledge_base=all` GET request gets all knowledge bases
`api/knowledgebase/<int:id>` PATCH request updates the object with whatever your payload contains, etc. etc

File: nextjsUIapp/next.config.mjs [hunk 10]
```diff
@@ -0,0 +1,11 @@
+/** @type {import('next').NextConfig} */
```
Comment (aaronkanzer) [inline nextjsUIapp/next.config.mjs:1]: Remove commented-out import?

File: nextjsUIapp/src/app/knowledge-base/page.tsx [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+                    ))}
+                </tr>
+                </thead>
+                <tbody>
+                {currentPageData.map((item, index) => (
+                    <tr key={index}
+                        className="bg-white border-b dark:bg-gray-800 dark:border-gray-700 hover:bg-gray-50 dark:hover:bg-gray-600">
+                        {headers.map((header, headerIndex) => (
+                            <td key={headerIndex}
+                                className="px-6 py-4">{item[header]?.value.substring(item[header]?.value.lastIndexOf('/') + 1)}</td>
```
Comment (aaronkanzer) [inline nextjsUIapp/src/app/knowledge-base/page.tsx:100]: Do you want a fallback value if item[header]? does not exist?

File: nextjsUIapp/src/app/knowledge-base/page.tsx [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+    const fetchData = async () => {
+        setLoading(true);
+        setError(null);
+        setData([]);
+        setHeaders([]);
+        setPageTitle("");
+        setSubPageTitle("");
+        const page = yaml.pages.find((page) => page.slug === "default");
+        const query_to_execute = page ? page.sparql_query : "";
+        console.log(query_to_execute);
```
Comment (aaronkanzer) [inline nextjsUIapp/src/app/knowledge-base/page.tsx:36]: Remove console.logs from function?

File: nextjsUIapp/src/app/knowledge-base/[slug]/page.tsx [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+                    ))}
+                </tr>
+                </thead>
+                <tbody>
+                {currentPageData.map((item, index) => (
+                    <tr key={index}
+                        className="bg-white border-b dark:bg-gray-800 dark:border-gray-700 hover:bg-gray-50 dark:hover:bg-gray-600">
+                        {headers.map((header, headerIndex) => (
+                            <td key={headerIndex}
+                                className="px-6 py-4">{item[header]?.value.substring(item[header]?.value.lastIndexOf('/') + 1)}</td>
```
Comment (aaronkanzer) [inline nextjsUIapp/src/app/knowledge-base/[slug]/page.tsx:99]: Do you want a fallback value if item[header]? does not exist?

File: nextjsUIapp/src/app/knowledge-base/[slug]/page.tsx [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+
+        try {
+            const response = await getData(queryParameter, endpoint, baseurl);
+            console.log('Raw response:', response);
+
+            if (response.status === 'success' && response.message?.results?.bindings) {
+                const bindings = response.message.results.bindings;
+                const vars = response.message.head.vars;
+                console.log('Bindings:', bindings);
+                console.log('Vars:', vars);
```
Comment (aaronkanzer) [inline nextjsUIapp/src/app/knowledge-base/[slug]/page.tsx:56]: Remove console.logs in this function?

File: nextjsUIapp/src/app/globals.css [hunk 15]
```diff
@@ -60,3 +60,27 @@ body {
...
+   margin:100px;
+}
+
+.fix-left-margin{
+    margin-left: 13rem;
+    margin-top:60px;
+}
+.kb-page-margin{
+    margin-top: 74px;
+    margin-right: 17px;
```
Comment (aaronkanzer) [inline nextjsUIapp/src/app/globals.css:73]: Use `rem` instead of px if possible

File: nextjsUIapp/src/app/data-release/page.tsx [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+                            NA
+                        </th>
+                        <td className="px-6 py-4">
+                            Na
+                        </td>
+                        <td className="px-6 py-4">
+                            NA
+                        </td>
+                        <td className="px-6 py-4">
+                            NA
```
Comment (aaronkanzer) [inline nextjsUIapp/src/app/data-release/page.tsx:71]: what do these `NA` values mean?

File: nextjsUIapp/src/app/knowledge-base/page.tsx (hunk 12 above)
Comment (tekrajchhetri) [reply nextjsUIapp/src/app/knowledge-base/page.tsx:36]: Sure I will remove all console.log

File: nextjsUIapp/src/app/data-release/page.tsx (hunk 16 above)
Comment (tekrajchhetri) [reply nextjsUIapp/src/app/data-release/page.tsx:71]: This is a skeleton page. Will be updated later.

File: nextjsUIapp/src/app/knowledge-base/page.tsx (hunk 11 above)
Comment (tekrajchhetri) [reply nextjsUIapp/src/app/knowledge-base/page.tsx]: > Do you want a fallback value if item[header]? does not exist?

No, at least not for the moment.

File: djangoUImanager/BrainKBUIManager/settings.py (hunk 2 above)
Comment (tekrajchhetri) [reply djangoUImanager/BrainKBUIManager/settings.py:26]: I have updated to read the key from env variable

File: djangoUImanager/Dockerfile (hunk 4 above)
Comment (tekrajchhetri) [reply djangoUImanager/Dockerfile:15]: Yes, it's intentional

File: djangoUImanager/README.md (hunk 3 above)
Comment (tekrajchhetri) [reply djangoUImanager/README.md:1]: now readme just contains the pointer to brainkbdocs

File: djangoUImanager/Web/models.py (hunk 8 above)
Comment (tekrajchhetri) [reply djangoUImanager/Web/models.py]: Updated model to "SET_NULL" on_delete

File: nextjsUIapp/next.config.mjs (hunk 10 above)
Comment (tekrajchhetri) [reply nextjsUIapp/next.config.mjs:1]: It's there by default but I've removed it now.

File: nextjsUIapp/src/app/knowledge-base/[slug]/page.tsx (hunk 14 above)
Comment (tekrajchhetri) [reply nextjsUIapp/src/app/knowledge-base/[slug]/page.tsx]: removed.

File: nextjsUIapp/src/app/globals.css (hunk 15 above)
Comment (tekrajchhetri) [reply nextjsUIapp/src/app/globals.css]: changed to rem as suggested

File: nextjsUIapp/src/app/page.tsx [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+
+    const [structuedModelHeaderTitle, setStructuedModelHeaderTitle] = useState("");
+    const [structuedModelHeaderSubTitle, setStructuedModelHeaderSubTitle] = useState("");
+
+    const [brainkbMainPageTitle, setBrainkbMainPageTitle] = useState("");
+    const [brainkbMainPageSubTitle, setBrainkbMainPageSubTitle] = useState("");
+
+    const [dataCount, setCountData] = useState<any[]>([])
+
+    const fetchAndSetData = async () => {
```
[the remark this answers is not in the document]
Comment (tekrajchhetri) [reply nextjsUIapp/src/app/page.tsx:19]: removed the console.log

File: djangoUImanager/Web/models.py (hunk 7 above)
Comment (tekrajchhetri) [reply djangoUImanager/Web/models.py]: updated to `on_delete=models.SET_NULL`

File: djangoUImanager/Web/models.py (hunk 6 above)
Comment (tekrajchhetri) [reply djangoUImanager/Web/models.py:59]: No reason, it's just I forget to add the id and have updated the model to include id in all case

**Comment by kabilar** [comment]
Hi @tekrajchhetri, great work on this project.  I am just catching up now.  When I try to log in with GitHub to http://3.149.235.227:3000/ I receive the error below.  Perhaps I am looking at the wrong deployment.

image

**Comment by tekrajchhetri** [comment]
> Hi @tekrajchhetri, great work on this project. I am just catching up now. When I try to log in with GitHub to http://3.149.235.227:3000/ I receive the error below. Perhaps I am looking at the wrong deployment.
> 
> image

@kabilar the deployment url is [https://beta.brainkb.org](https://beta.brainkb.org). I am not sure how to disable IP based access. Also, if you have any comments/suggestion for the following issue below that would be great.

image

Even after enabling https on all URL, I am still getting this issue.

**Comment by kabilar** [comment]
Thanks for the quick response, @tekrajchhetri.  I will file these as separate issues so its easier to discuss each in a separate thread.

File: nextjsUIapp/src/app/about/page.tsx [hunk 18]
```diff
@@ -1,13 +1,19 @@
...
+export default async function About(){
+        const current_session = await getServerSession(authOptions);
+    //user is logged in so redirect to admin page
+    if (current_session) return redirect("/admin");
     return (
-        <div>
+        <div className="set-margin-hundred">
 
             <div className="text-left">
                 <h2 className="mb-4 text-3xl font-extrabold font-extrabold leading-none text-sky-900 animate-slide-up">What is BrainKB?</h2>
```
Comment (kabilar) [inline nextjsUIapp/src/app/about/page.tsx:19]:
```suggestion
                <h2 className="mb-4 text-3xl font-extrabold font-extrabold leading-none text-white animate-slide-up">What is BrainKB?</h2>
```

Comment (kabilar) [reply nextjsUIapp/src/app/about/page.tsx:19]: I would suggest changing all text to white since `sky-900` is hard to read on a blue background.

image

Comment (tekrajchhetri) [reply nextjsUIapp/src/app/about/page.tsx:19]: @kabilar It's little bit strange because I do not see blue background.

This is what I see.

image

Comment (kabilar) [reply nextjsUIapp/src/app/about/page.tsx:19]: It looks like the site is responsive to the theme of the operating system (i.e. light or dark mode).  When I change to light mode, I see the same color scheme in the screenshot that you provided above.  Perhaps we need to fix the background to be white regardless of the theme.  Or change the text to be responsive to the theme.

**Comment by tekrajchhetri** [comment]
I merged this request and will create a small one similar to BrainKB.
